modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/serviceextensions/preupdatepassword/PreUpdatePasswordActionSmsOtpFailureTestCase.java

```diff
@@ -0,0 +1,290 @@
+import org.wso2.identity.integration.test.recovery.model.v2.UserClaim;
+import org.wso2.identity.integration.test.rest.api.server.identity.governance.v1.dto.ConnectorsPatchReq;
+import org.wso2.identity.integration.test.rest.api.server.identity.governance.v1.dto.PropertyReq;
+import org.wso2.identity.integration.test.rest.api.server.notification.sender.v1.model.Properties;
+import org.wso2.identity.integration.test.rest.api.server.notification.sender.v1.model.SMSSender;
+import org.wso2.identity.integration.test.restclients.NotificationSenderRestClient;
+import org.wso2.identity.integration.test.restclients.PasswordRecoveryV2RestClient;
+import org.wso2.identity.integration.test.serviceextensions.dataprovider.model.ActionResponse;
+import org.wso2.identity.integration.test.serviceextensions.dataprovider.model.ExpectedPasswordUpdateResponse;
+import org.wso2.identity.integration.test.serviceextensions.mockservices.ServiceExtensionMockServer;
+import org.wso2.identity.integration.test.serviceextensions.model.*;
+import org.wso2.identity.integration.test.rest.api.server.notification.sender.v1.model.Error;
+import org.wso2.identity.integration.test.rest.api.user.common.model.*;
+import org.wso2.identity.integration.test.restclients.SCIM2RestClient;
+import org.wso2.identity.integration.test.utils.FileUtils;
+
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.servlet.http.HttpServletResponse;
+
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNotNull;
+import static org.testng.Assert.assertTrue;
+
+/**
+ * Integration test class for testing the pre update password action failure.
+ * This test case extends {@link PreUpdatePasswordActionBaseTestCase} and focuses on failed scenario
+ * on SMS OTP based password recovery/reset via V2 recovery API.
+ */
+public class PreUpdatePasswordActionSmsOtpFailureTestCase extends PreUpdatePasswordActionBaseTestCase {
+
+    private static final String MOBILE = "+941111111111";
+    private static final String SMS_SENDER_REQUEST_FORMAT = "{\"content\": {{body}}, \"to\": {{mobile}} }";
+
+    private final String tenantId;
+    private final TestUserMode userMode;
+
+    private SCIM2RestClient scim2RestClient;
+    private NotificationSenderRestClient notificationSenderRestClient;
+    private PasswordRecoveryV2RestClient passwordRecoveryV2RestClient;
+
+    private String actionId;
+    private String userId;
+    private UserObject userInfo;
+
+    private ServiceExtensionMockServer serviceExtensionMockServer;
+    private MockSMSProvider mockSMSProvider;
+
+    private final ActionResponse actionResponse;
+    private final ExpectedPasswordUpdateResponse expectedPasswordUpdateResponse;
+
+    @Factory(dataProvider = "testExecutionContextProvider")
+    public PreUpdatePasswordActionSmsOtpFailureTestCase(TestUserMode testUserMode, ActionResponse actionResponse,
```

Comment:
Also since we have all the base methods for sms otp scenario, shall we add a test for success case well.

Reply:
Lets improve this in another PR immediately.